### ISSUE TITLE
how to implement softer-NMS by pytorch version

### ISSUE THREAD

**Issue by gittigxuy** [yihui-he/softer-NMS#8]
just change a little bit code in soft NMS ？could you please illustrate detailly?

**Comment by LiDaiY** [comment]
I saw the source code，it seems we should predict the standard variance before executing NMS. So there should be more change?

**Comment by JoyHuYY1412** [comment]
I implemented only the KL-loss part in pytorch.

I haven't test but you can check dies it work. The bbox_pred_std is the 'alpha' in the paper.

 ```
   def smooth_l1_loss(self, bbox_pred, bbox_targets, bbox_inside_weights, bbox_outside_weights, sigma=1.0):    
        sigma_2 = sigma ** 2
        box_diff = bbox_pred - bbox_targets
        in_box_diff = bbox_inside_weights * box_diff
        abs_in_box_diff = torch.abs(in_box_diff)
        smoothL1_sign = (abs_in_box_diff < 1. / sigma_2).detach().float()
        loss_box = (torch.pow(in_box_diff, 2) * (sigma_2 / 2.) * smoothL1_sign
                    + (abs_in_box_diff - (0.5 / sigma_2)) * (1. - smoothL1_sign)) * bbox_outside_weights
        return loss_box.sum() / loss_box.shape[0]

    def KL_loss(self, bbox_pred, bbox_targets, bbox_pred_std, bbox_inside_weights, bbox_outside_weights, sigma=1.0):
        #KL-loss
        sigma_2 = sigma ** 2
        box_diff = bbox_pred - bbox_targets
        in_box_diff = bbox_inside_weights * box_diff #bbox_inw = in_box_diff
        bbox_l1abs = torch.abs(in_box_diff)  #abs_in_box_diff = bbox_l1abs
        # bbox_sq = in_box_diff * in_box_diff
        smoothL1_sign = (bbox_l1abs < 1. / sigma_2).detach().float()  #1 if bbox_l1abs<1 else 0 
        bbox_inws = (torch.pow(in_box_diff, 2) * (sigma_2 / 2.) * smoothL1_sign
                    + (bbox_l1abs - (0.5 / sigma_2)) * (1. - smoothL1_sign)) 
        bbox_inws = bbox_inws.detach().float()  #?? to be confirmed
        scale = 1
        bbox_pred_std_abs_log = bbox_pred_std*0.5*scale
        bbox_pred_std_nabs = -1*bbox_pred_std
        bbox_pred_std_nexp = torch.exp(bbox_pred_std_nabs)
        bbox_inws_out = bbox_pred_std_nexp * bbox_inws
        bbox_pred_std_abs_logw = bbox_pred_std_abs_log * bbox_outside_weights
        bbox_pred_std_abs_logwr = torch.mean(bbox_pred_std_abs_logw, dim = 0)
        
        #bbox_pred grad, stop std
        loss_bbox = self.smooth_l1_loss(bbox_pred, bbox_targets, bbox_inside_weights, bbox_pred_std_nexp)
        bbox_pred_std_abs_logw_loss = torch.sum(bbox_pred_std_abs_logwr)
        bbox_inws_out = bbox_inws_out * scale
        bbox_inws_outr = torch.mean(bbox_inws_out, dim = 0)
        bbox_pred_std_abs_mulw_loss = torch.sum(bbox_inws_outr)
        return (loss_bbox + bbox_pred_std_abs_logw_loss + bbox_pred_std_abs_mulw_loss)
```

**Comment by wysot** [comment]
@JoyHuYY1412 
What are the bbox_inside_weights and bbox_outside_weights in the KL_LOSS function?

**Comment by yihui-he** [comment]
3rd party reimplementation in PyTorch: https://github.com/yihui-he/KL-Loss/issues/20
BTW, this repo is deprecated. Please use our new repo: https://github.com/yihui-he/KL-Loss
